### PR TITLE
ros-noetic driver for serow + small changes

### PR DESCRIPTION
**Pull request by MichaelMarav**
### Description
* Added a ros driver for serow (single node) subs to topics and calls the serow filter. Publishes the base state under /serow/base_estimate
* added the remapper for the valk .bag that corrects the frames (.bag is to large to push) 
* added valk json (non tuned) 
* Removed redundant if (ft.has_value) from serow.cpp (if it doesn't the function returns at the the start)
* Added check for correct frames in the wrench measurement
* added vs code, devel and build folders to gitignore
* split ros1 and ros2 into two workspaces: ros_ws and ros2_ws

### CODE REVIEW

File: core/src/Serow.cpp [hunk 1]
```diff
@@ -262,72 +271,74 @@ void Serow::filter(ImuMeasurement imu, std::map<std::string, JointMeasurement> j
             kinematic_estimator_->linearVelocity(contact_frame);
     }
 
-    if (ft.has_value()) {
```
Comment (mrsp) [inline core/src/Serow.cpp:265]: revert

File: ros_ws/.catkin_workspace [hunk 2]
```diff
@@ -0,0 +1 @@
+# This file currently only serves to mark the location of a catkin workspace for tool integration
```
Comment (mrsp) [inline ros_ws/.catkin_workspace:1]: remove

File: ros_ws/src/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+/opt/ros/noetic/share/catkin/cmake/toplevel.cmake
```
Comment (mrsp) [inline ros_ws/src/CMakeLists.txt:1]: ?

File: ros_ws/src/serow_ros/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+add_executable(serow_ros src/serow_driver.cpp)
+target_link_libraries(serow_ros ${catkin_LIBRARIES} Eigen3::Eigen pinocchio::pinocchio serow)
+add_dependencies(serow_ros ${catkin_EXPORTED_TARGETS})
+
+set_property(TARGET serow_ros PROPERTY CXX_STANDARD 17)
+
+
+catkin_install_python(PROGRAMS src/force_remapper.py
+  DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION}
+)
```
Comment (mrsp) [inline ros_ws/src/serow_ros/CMakeLists.txt:37]: add empty line

File: ros_ws/src/serow_ros/package.xml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  <maintainer email="michael@todo.todo">michael</maintainer>
+
+
+  <!-- One license tag required, multiple allowed, one license per tag -->
+  <!-- Commonly used license strings: -->
+  <!--   BSD, MIT, Boost Software License, GPLv2, GPLv3, LGPLv2.1, LGPLv3 -->
+  <license>TODO</license>
+
+
+  <!-- Url tags are optional, but multiple are allowed, one per tag -->
```
Comment (mrsp) [inline ros_ws/src/serow_ros/package.xml:19]: remove commented code add license

File: ros_ws/src/serow_ros/src/force_remapper.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env python
+
+import rospy
```
Comment (mrsp) [inline ros_ws/src/serow_ros/src/force_remapper.py:3]: why is this need to be in the repo?

File: ros_ws/src/serow_ros/src/serow_driver.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    ros::Subscriber base_imu_subscription_;
+    ros::Publisher base_state_publisher_ ;
+    
+    std::vector<ros::Subscriber> force_torque_state_subscriptions_;
+    std::queue<sensor_msgs::JointState> joint_state_data_;
+    std::queue<sensor_msgs::Imu> base_imu_data_;
+    std::map<std::string, std::queue<geometry_msgs::WrenchStamped>> ft_data_;
+    geometry_msgs::PoseStamped pose_estimate_;
+    serow::Serow serow_;
+    // serow::State state_;
```
Comment (mrsp) [inline ros_ws/src/serow_ros/src/serow_driver.cpp:143]: remove commented code

File: ros_ws/src/serow_ros/src/serow_driver.cpp [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+};
+
+int main(int argc, char** argv) {
+    ros::init(argc, argv, "serow_ros_driver");
+    ros::NodeHandle nh;
+
+    std::vector<std::string> force_torque_state_topics;
+    force_torque_state_topics.push_back("/ihmc_ros/valkyrie/output/foot_force/left");
```
Comment (mrsp) [inline ros_ws/src/serow_ros/src/serow_driver.cpp:152]: parametrize these

**Review by mrsp — COMMENTED** [review]
Let's have folders ros and ros2 instead of ros_ws and ros2_ws

File: ros_ws/src/serow_ros/src/serow_driver.cpp [hunk 9]
```diff
@@ -0,0 +1,159 @@
+#include "pinocchio/fwd.hpp"
+#include <functional>
+#include <map>
+#include <queue>
```
Comment (mrsp) [inline ros_ws/src/serow_ros/src/serow_driver.cpp:4]: alphabetical order

File: ros/src/serow_ros/src/serow_driver.cpp [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+                                                ft_data.wrench.force.z);
+                        ft.torque = Eigen::Vector3d(ft_data.wrench.torque.x,
+                                                    ft_data.wrench.torque.y,
+                                                    ft_data.wrench.torque.z);
+                        ft_measurements[key] = std::move(ft);
+                        value.pop();
+                    }
+                }
+
+                serow_.filter(imu_measurement, joint_measurements,
```
Comment (mrsp) [inline ros/src/serow_ros/src/serow_driver.cpp:121]: nice

File: ros/src/serow_ros/src/serow_driver.cpp [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+                auto state_ = serow_.getState();
+                pose_estimate_.header.seq  += 1;
+                pose_estimate_.pose.position.x = state_->getBasePosition().x();
+                pose_estimate_.pose.position.y = state_->getBasePosition().y();
+                pose_estimate_.pose.position.z = state_->getBasePosition().z();
+                pose_estimate_.pose.orientation.x = state_->getBaseOrientation().coeffs().x();
+                pose_estimate_.pose.orientation.y = state_->getBaseOrientation().coeffs().y();
+                pose_estimate_.pose.orientation.z = state_->getBaseOrientation().coeffs().z();
+                pose_estimate_.pose.orientation.w = state_->getBaseOrientation().coeffs().w();
+                base_state_publisher_.publish(pose_estimate_);
```
Comment (mrsp) [inline ros/src/serow_ros/src/serow_driver.cpp:134]: probably we want to publish base velocity and com position + velocity as well

File: ros/src/serow_ros/src/serow_driver.cpp [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+        // Initialize SERoW
+        serow_ = serow::Serow(config_file_path_);
+        pose_estimate_.header.frame_id = "base_link";
+
+        // Create subscribers
+        joint_state_subscription_ = nh_.subscribe(joint_state_topic_, 1000, &SerowDriver::joint_state_topic_callback, this);
+        base_imu_subscription_ = nh_.subscribe(base_imu_topic_, 1000, &SerowDriver::base_imu_topic_callback, this);
+        base_state_publisher_ = nh_.advertise<geometry_msgs::PoseStamped>("/serow/base_estimate", 100);
```
Comment (mrsp) [inline ros/src/serow_ros/src/serow_driver.cpp:37]: call this /serow/base/odom and use /serow/base/twist (up for discussion) for velocity 
similarly for com /serow/com/odom and /serow/com/twist

**Review by mrsp — APPROVED** [review]
LGTM

File: ros/src/serow_ros/src/serow_driver.cpp (hunk 11 above)
Comment (MichaelMarav) [reply ros/src/serow_ros/src/serow_driver.cpp:134]: Will add it after we make sure that the base estimates are fixed

**Pull request merged**